syntax = "proto3";

message CoreStat {
    optional uint64 user       = 1;
    optional uint64 nice       = 2;
    optional uint64 system     = 3;
    optional uint64 idle       = 4;
    optional uint64 iowait     = 5;
    optional uint64 irq        = 6;
    optional uint64 softirq    = 7;
    optional uint64 steal      = 8;
    optional uint64 guest      = 9;
    optional uint64 guest_nice = 10;
}

message KernelMiscStat {
    optional uint64 intr          = 1;
    optional uint64 ctxt          = 2;
    optional uint64 processes     = 3;
    optional uint32 procs_running = 4;
    optional uint32 procs_blocked = 5;
}

message SoftIRQStat {
    optional uint64 total    = 1;
    optional uint64 hi       = 2;
    optional uint64 timer    = 3;
    optional uint64 net_tx   = 4;
    optional uint64 net_rx   = 5;
    optional uint64 block    = 6;
    optional uint64 irq_poll = 7;
    optional uint64 tasklet  = 8;
    optional uint64 sched    = 9;
    optional uint64 hrtimer  = 10;
    optional uint64 rcu      = 11;
}

// Refer to https://man7.org/linux/man-pages/man5/proc_stat.5.html
message CPUMetrics {
    optional uint64          timestamp        = 1;
    repeated CoreStat        core_stats       = 2;
    optional KernelMiscStat  kernel_misc_stat = 3;
    optional SoftIRQStat     soft_irq_stat    = 4;
}

message CPUMetricsTimeSeries {
    repeated CPUMetrics metrics = 1;
}
